lib/task-queue/TaskRouter.js

```diff
@@ -0,0 +1,132 @@
+/* globals Task */
+"use strict";
+
+const {Cu} = require("chrome");
+const {WorkerTask} = require("lib/task-queue/WorkerTask");
+const {Storage} = require("lib/task-queue/Storage");
+const {Page} = require("lib/task-queue/Page");
+
+Cu.import("resource://gre/modules/Task.jsm");
+
+// The lock for page and task creation test-and-set
+let lock = false;
+// The queue for messages waiting to be processed
+const mutexQueue = [];
+
+class TaskRouter {
+  /**
+   * Creates a new task router from a map of messages to TaskProcessor arrays
+   *
+   * @param {Map<string, TaskProcessor[]>} processorMap
+   *                                       A map of Task Processors arrays for each message
+   */
+  constructor(processorMap) {
+    if (!processorMap) {
+      throw new Error("TaskRouter needs a map as argument.");
+    }
+    // Initialize task processors
+    this.routes = processorMap;
+  }
+
+  /**
+   * Check if the page should be processed by the queue
+   *
+   * @param {String}    url
+   *                    The URL of the page we want to check
+   * @returns {Promise}
+   *                    Returns true if the page should be processed, false otherwise
+   */
+  asyncNeedProcessing(url) {
+    return Task.spawn((function*() {
+      // Do we already have a task for this url?
+      const tasks = yield WorkerTask.asyncGetByUrl(url);
+      if (tasks) {
+        return false;
+      }
+
+      // Do we already have a page for this url?
+      const page = yield Page.asyncGetByUrl(url);
+      if (!page) {
+        return true;
+      }
+
+      // Is the page still valid?
+      if (page.expiration > Date.now()) {
+        return false;
+      }
+
+      // Do we have a DOM file for this page?
+      const file = yield Storage.asyncGetFile(page.path);
+      if (!file) {
+        return false;
+      }
+
+      return true;
+    }).bind(this));
+  }
+
+  /**
+   * Create the worker tasks for a given message
+   *
+   * @param {Object} message
+   *                 The message we want to create the tasks for
+   */
+  asyncCreateTasks(message) {
+    return Task.spawn((function*() {
+      lock = true;
```

Comment:
You need to acquire the lock. meaning, the value of the lock needs to be `false` before the code proceeds.
Otherwise, there is no mutual exclusion